### PR TITLE
Update generator for Unicode 12.1

### LINKED ISSUE #344: Add facebook and whatsapp emojis
Please add facebook and whatsapp emojis too.
Thanks

----------------------------------------------------------------------

### PR DESCRIPTION
**Pull request by rubengees**
This PR is a bigger overhaul of the generator to support the latest changes to our data source among other things:

- Update to latest emojis and make necessary changes to the generator.
- Add google compat to the targets (only code generated).
- Fix low(er) quality images.
- Add facebook module.
- Update dependencies.

This comes with a significant change how we generate the categories. Due to too many icons being present now we started to run into "code too large" exceptions. I have changed the generator so that it generates "chunk" classes of 250 emojis at most and concats them together in the actual category.

Before:

```
category
-- PeopleCategory
-- SymbolsCategory
-- TravelCategory
```

Now:

```
category
-- SmileysAndPeopleCategory
-- SmileysAndPeopleCategoryChunk0
-- SmileysAndPeopleCategoryChunk1
-- SymbolsCategory
-- SymbolsCategoryChunk0
-- TravelAndPlacesCategory
-- TravelAndPlacesCategoryChunk0
-- CategoryUtils
```

As always this PR comes without the generated code to make review easier.
Would close #344 and #356.

### CODE REVIEW

File: generator/index.js [hunk 1]
```diff
@@ -70,43 +100,62 @@ async function copyTargetImages(map, target, shouldOptimize) {
...
     });
 
-    const src = `node_modules/emoji-datasource-${target.dataSource}/img/${target.dataSource}/sheets/64.png`;
-    const sheet = await Jimp.read(src);
-    const strips = sheet.bitmap.width / 66 - 1;
-    for (let i = 0; i < strips; i++) {
-        const maxY = emojiByStrip[i].map(it => it.sheet_y).reduce((a, b) => Math.max(a, b), 0);
-        const height = (maxY + 1) * 66;
-        const strip = await new Jimp(66, height);
-        await strip.blit(sheet, 0, 0, i * 66, 0, 66, height);
```
Comment (rubengees) [inline generator/index.js]: This was the cause of lower quality images.

File: generator/index.js [hunk 2]
```diff
@@ -70,43 +100,62 @@ async function copyTargetImages(map, target, shouldOptimize) {
...
+        const src = `node_modules/emoji-datasource-${target.dataSource}/img/${target.dataSource}/sheets-clean/64.png`;
+        const sheet = await Jimp.read(src);
+        const strips = sheet.bitmap.width / 66 - 1;
+
+        for (let i = 0; i < strips; i++) {
+            const dest = `../emoji-${target.module}/src/main/res/drawable-nodpi/emoji_${target.module}_sheet_${i}.png`;
+            const maxY = emojiByStrip[i].map(it => it.sheet_y).reduce((a, b) => Math.max(a, b), 0);
+            const height = (maxY + 1) * 66;
+
+            const strip = await sheet.clone().crop(i * 66, 0, 66, height)
```
Comment (rubengees) [inline generator/index.js]: We now simply crop.

File: generator/index.js [hunk 3]
```diff
@@ -189,6 +248,14 @@ async function parse() {
         }
     }
 
+    // Normalize the new "Smileys & Emotion" and "People & Body" categories to the ones we have.
+    const smileysAndEmotion = result.get("SmileysAndEmotion")
+    const peopleAndBody = result.get("PeopleAndBody")
+
+    result.set("SmileysAndPeople", smileysAndEmotion.concat(peopleAndBody))
+    result.delete("SmileysAndEmotion")
+    result.delete("PeopleAndBody")
```
Comment (rubengees) [inline generator/index.js]: Our upstream source slightly changed the categories. Here we normalize them back to what we had previously.

File: generator/index.js [hunk 4]
```diff
@@ -260,17 +365,33 @@ async function generateCode(map, targets) {
...
+        if (target.module !== "google-compat") {
+            await fs.writeFile(`${srcDir}/${target.name}.java`, template(emojiTemplate)({
+                package: target.package,
+                name: target.name,
+                strips: strips,
+            }));
+        }
 
-        await fs.writeFile(`${dir}/${target.name}.java`, _(emojiTemplate).template()({
+        await fs.writeFile(`${valuesDir}/strings.xml`, template(stringsTemplate)({
```
Comment (rubengees) [inline generator/index.js]: We now also generate the `strings.xml` file automatically.

**Comment by rubengees** [comment]
I have added the generated files and will merge once CI passes and @mario approves :slightly_smiling_face:.

**Comment by mario** [comment]
Approved by Mario (tm)

**Pull request merged**